lib/ContextResolver.js

```diff
@@ -0,0 +1,243 @@
+      }
+      if(ctx === null) {
+        // handle `null` context, nothing to cache
+        allResolved.push(new ResolvedContext({document: null}));
+        continue;
+      }
+      if(!_isObject(ctx)) {
+        _throwInvalidLocalContext(context);
+      }
+      // context is an object, get/create `ResolvedContext` for it
+      const key = JSON.stringify(ctx);
+      let resolved = this._get(key);
+      if(!resolved) {
+        // create a new static `ResolvedContext` and cache it
+        resolved = new ResolvedContext({document: ctx});
+        this._cacheResolvedContext({key, resolved, tag: 'static'});
+      }
+      allResolved.push(resolved);
+    }
+
+    return allResolved;
+  }
+
+  _get(key) {
+    // get key from per operation cache; no `tag` is used with this cache so
+    // any retrieved context will always be the same during a single operation
+    let resolved = this.perOpCache.get(key);
+    if(!resolved) {
+      // see if the shared cache has a `static` entry for this URL
+      const tagMap = this.sharedCache.get(key);
+      if(tagMap) {
+        resolved = tagMap.get('static');
+        if(resolved) {
+          this.perOpCache.set(key, resolved);
+        }
+      }
+    }
+    return resolved;
+  }
+
+  _cacheResolvedContext({key, resolved, tag}) {
+    this.perOpCache.set(key, resolved);
+    if(tag !== undefined) {
+      let tagMap = this.sharedCache.get(key);
+      if(!tagMap) {
+        tagMap = new Map();
+        this.sharedCache.set(key, tagMap);
+      }
+      tagMap.set(tag, resolved);
+    }
+    return resolved;
+  }
+
+  async _resolveRemoteContext({url, documentLoader, base, cycles}) {
+    // resolve relative URL and fetch context
+    url = prependBase(base, url);
+    const {context, remoteDoc} = await this._fetchContext(
+      {url, documentLoader, cycles});
+
+    // update base according to remote document and resolve any relative URLs
+    base = remoteDoc.documentUrl || url;
+    _resolveContextUrls({context, base});
+
+    // resolve, cache, and return context
+    const resolved = await this.resolve(
+      {context, documentLoader, base, cycles});
+    this._cacheResolvedContext({key: url, resolved, tag: remoteDoc.tag});
+    return resolved;
+  }
+
+  async _fetchContext({url, documentLoader, cycles}) {
+    // check for max context URLs fetched during a resolve operation
+    if(cycles.size > MAX_CONTEXT_URLS) {
+      throw new JsonLdError(
+        'Maximum number of @context URLs exceeded.',
+        'jsonld.ContextUrlError',
+        {code: 'loading remote context failed', max: MAX_CONTEXT_URLS});
+    }
+
+    // check for context URL cycle
+    if(cycles.has(url)) {
+      throw new JsonLdError(
+        'Cyclical @context URLs detected.',
+        'jsonld.ContextUrlError',
+        {code: 'recursive context inclusion', url});
+    }
+
+    // track cycles
+    cycles.add(url);
+
+    let context;
+    let remoteDoc;
+
+    try {
+      remoteDoc = await documentLoader(url);
```

Comment:
Note that we're going to need to pass in API arguments to the documentLoader for options such as `extractAllScripts`, but this can be handled in a different PR.

Reply:
Yeah. I didn't want to do anything with that in this PR but it will need to be addressed in a subsequent PR.